src/ipvs/ip_vs_proto_udp.c

```diff
@@ -189,59 +200,100 @@ static int send_standalone_uoa(const struct dp_vs_conn *conn,
+        memcpy(&((struct ip6_hdr *)iph)->ip6_src,
+                                      &((struct ip6_hdr *)oiph)->ip6_src,
+                                      IPV6_ADDR_LEN_IN_BYTES);
+        memcpy(&((struct ip6_hdr *)iph)->ip6_dst,
+                                      &((struct ip6_hdr *)oiph)->ip6_dst,
+                                      IPV6_ADDR_LEN_IN_BYTES);
+    } else {
+        iph = (void *)rte_pktmbuf_append(mbuf, sizeof(struct iphdr));
+        if (unlikely(!iph))
+            goto no_room;
+        ((struct iphdr *)iph)->version = 4;
+        ((struct iphdr *)iph)->tos     = ((struct iphdr *)oiph)->tos;
+        ((struct iphdr *)iph)->id      = ip4_select_id((struct ipv4_hdr *)iph);
+        ((struct iphdr *)iph)->frag_off = 0;
+        ((struct iphdr *)iph)->ttl     = ((struct iphdr *)oiph)->ttl;
+        ((struct iphdr *)iph)->saddr   = conn->laddr.in.s_addr;
+        ((struct iphdr *)iph)->daddr   = conn->daddr.in.s_addr;
+    }
 
     if (mode == UOA_M_IPO) {
-        iph->ihl    = (sizeof(struct iphdr) + IPOLEN_UOA) / 4;
-        iph->tot_len = htons(sizeof(*iph) + IPOLEN_UOA + sizeof(*uh));
-        iph->protocol = oiph->protocol; /* should always UDP */
-
-        uoa = (void *)rte_pktmbuf_append(mbuf, IPOLEN_UOA);
+        /* only ipv4 support and use this ip option mode */
+        ((struct iphdr *)iph)->ihl =
+            (sizeof(struct iphdr) + IPOLEN_UOA_IPV4) / 4;
+        ((struct iphdr *)iph)->tot_len  =
+            htons(sizeof(*iph) + IPOLEN_UOA_IPV4 + sizeof(*uh));
+        ((struct iphdr *)iph)->protocol = ((struct iphdr *)oiph)->protocol;
+
+        uoa = (void *)rte_pktmbuf_append(mbuf, ipolen_uoa);
     } else { /* UOA_M_OPP */
-        iph->ihl    = sizeof(struct iphdr) / 4;
-        iph->tot_len = \
-            htons(sizeof(*iph) + sizeof(*opp) + sizeof(*uoa) + sizeof(*uh));
-        iph->protocol = IPPROTO_OPT;
+        if (AF_INET6 == af) {
+            ((struct ip6_hdr *)iph)->ip6_plen =
+                            /* htons(ntohs(((struct ip6_hdr *)oiph)->ip6_plen) + */
+                            sizeof(*opp) + sizeof(*uoa) + sizeof(*uh);
+            ((struct ip6_hdr *)iph)->ip6_nxt = IPPROTO_OPT;
+        } else {
+            ((struct iphdr *)iph)->ihl = sizeof(struct iphdr) / 4;
+            ((struct iphdr *)iph)->tot_len = htons(sizeof(struct iphdr) +
+                                sizeof(*opp) + sizeof(*uoa) + sizeof(*uh));
+            ((struct iphdr *)iph)->protocol = IPPROTO_OPT;
+        }
 
         /* option-proto */
         opp = (void *)rte_pktmbuf_append(mbuf, sizeof(*opp));
         if (unlikely(!opp))
             goto no_room;
 
         memset(opp, 0, sizeof(*opp));
-        opp->version = 0x01;
-        opp->protocol = oiph->protocol;
-        opp->length = htons(sizeof(*opp) + sizeof(*uoa));
+        if (AF_INET6 == af) {
+            opp->version  = OPPHDR_IPV6;
+            opp->protocol = ((struct ip6_hdr *)oiph)->ip6_nxt;
```

Comment:
Is `oiph->ip6_nxt`  IPPROTO_UDP when exthdrs exist?

Reply:
when in send_standalone_uoa, set to fixed IPPROTO_UDP